Add support for checking tensor containers in `torch.testing`

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56373 add `torch.testing` to docs
* #56365 Split `test_testing.py::TestAsserts` for multiple devices
* #55890 enable support for custom error messages in `torch.testing`
* #55788 add support for equal_nan in torch.testing.assert_close
* #55786 Add support for scalars and numpy in torch.testing
* **#55385 Add support for checking tensor containers in `torch.testing`**

This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

Differential Revision: [D27903805](https://our.internmc.facebook.com/intern/diff/D27903805)

pmeier committed:
Add support for checking tensor containers in `torch.testing`

This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

mruberry commented:
Looks like this PR in the stack is just a single blank line?

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -331,29 +368,230 @@ def assert_tensors_equal(
+E = TypeVar("E", bound=Exception)
+
+
+def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
```
I wonder if the need for this function already warrants a change from all internal checks returning `exc: Optional[Exception]` to `exc_type: Optional[Type[Exception]], msg: Optional[str]`. It would make the default case a little more verbose

```python
exc_type, msg = _check_whatever()
if exc_type:
    raise exc_type(msg)
```

vs.

```python
exc = _check_whatever()
if exc:
    raise exc
```
but remove this awkward re-instantiation of the exception.

mruberry replied:
Either seems reasonable for now; I like being verbose

pmeier replied:
Lets keep it they way it is right now and think about it after more pressing issues in the stack are resolved.

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

mruberry commented on test/test_testing.py:
```diff
@@ -740,36 +747,103 @@ def test_trivial_passing_test(self, device):
+        ]
+
+    def assert_fns_with_inputs(self, actual: torch.Tensor, expected: torch.Tensor) -> Iterator[Callable]:
+        """Yields assert functions with with included positional inputs based on two example tensors.
```
"with with" -> "with"

mruberry commented on test/test_testing.py:
```diff
@@ -740,36 +747,103 @@ def test_trivial_passing_test(self, device):
+            yield functools.partial(assert_fn, *inputs)
+
+    @contextlib.contextmanager
+    def assertRaisesRegexs(self, expected_exception: Type[Exception], *expected_regexs: str) -> Iterator[None]:
```
Don't we already have this context manager in the test suite?

pmeier replied:
I don't think so. At least I didn't find it. This is basically an upgrade to the builtin `assertRaisesRegex` (mind the missing "s" at the end) that accepts multiple regexs. If you want I can move it to our `TestCase`.

Without something like that it is really hard to check if an error message contains multiple unordered parts. For an example see `test_unknown_type` (L1013-L1020): We want to make sure that the type of `actual` and `expected` appear in the message, but it doesn't matter which comes first.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -331,29 +368,230 @@ def assert_tensors_equal(
-    actual: torch.Tensor,
-    expected: torch.Tensor,
+def assert_close(
+    actual: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
```
So we won't support lists of dicts whose values are lists containing tensors?

pmeier replied:
Correct. Only top level containers are supported. Anything else will be a lot more involved to handle and I honestly can't imagine a wide-spread scenario that would warrant this.

pmeier commented on test/test_testing.py:
```diff
@@ -778,220 +852,214 @@ def test_quantized_support(self, device):
+        expected[0, 1] = 2
+        expected[1, 2] = 9
+
+        expected_regexs = (
+            r"\s+2\s+",  # absolute number of mismatches
+            r"22([.]2+)?\s*[%]",  # relative number of mismatches
+            r"\s+4[.]0\s+",  # maximum absolute difference
+            r"1,\s*2",  # index of maximum absolute difference
+            r"\s+0[.]5\s+",  # maximum relative difference
+            r"0,\s*1",  # index of maximum relative difference
+        )
```
After some offline discussion @mattip recommended to not use a regex or worse multiple regexs to check an error message to make the test more readable. I've introduced the regex checking to allow the error message some flexibility in the wording and only check the numbers. Opinion @mruberry?

mruberry replied:
I would also prefer to avoid adding a new regular expression assertion context manager.

One option would be to simplify this test by making it less programmatic and just asserting string equality with an expected message.

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

Differential Revision: [D27903805](https://our.internmc.facebook.com/intern/diff/D27903805)

[ghstack-poisoned]

mruberry commented:
Hey @pmeier, unfortunately I can't land this because I can't import it using ghimport. ghimport suggests that a rebase might help, though -- would you rebase this stack and ping me?

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

Differential Revision: [D27903805](https://our.internmc.facebook.com/intern/diff/D27903805)

[ghstack-poisoned]

pmeier commented:
No merge conflicts, so not sure what happened.

pmeier committed:
Update on "Add support for checking tensor containers in `torch.testing`"


This renames `assert_tensors_(equal|close)` to `_check_tensors_(equal|close)` and exposes two new functions: `assert_(equal|close)`. In addition to tensor pairs, the newly added functions also support the comparison of tensors in sequences or mappings. Otherwise their signature stays the same.

Differential Revision: [D27903805](https://our.internmc.facebook.com/intern/diff/D27903805)

[ghstack-poisoned]
